### PR TITLE
[TRTLLM-4638	][feat] add best of n support with reward model in scaffolding

### PR DESCRIPTION
**Pull request by dc3671**
# PR title

Please write the PR title by following template:

[JIRA ticket link/nvbug link/github issue link][fix/feat/doc/infra/...] \<summary of this PR\>

For example, assume I have a PR hope to support a new feature about cache manager of Jira TRTLLM-1000 ticket, it would be like

[TRTLLM-1000][feat] Support a new feature about cache manager

## Description

Please explain the issue and the solution in short.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: examples/scaffolding/run_majority_vote_aime24.py
Comment (WeiHaocheng) [inline examples/scaffolding/run_majority_vote_aime24.py:1]: Maybe name it "run_majority_vote.py"? This looks more symmetrical.

File: examples/scaffolding/run_best_of_n_with_reward.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    })
+    reward_controller = QwenRewardController(tokenizer=reward_worker.tokenizer)
+    controller = BestOfNController(
+        generation_controller=gen_controller,
+        reward_controller=reward_controller,
+        default_sample_num=args.sample_num,
+    )
+    llm = ScaffoldingLlm(controller, workers=workers)
+
+    # query = "Solve for x: 4x + 5 = 6x - 7"
```
Comment (WeiHaocheng) [inline examples/scaffolding/run_best_of_n_with_reward.py:58]: debugging information

File: tensorrt_llm/scaffolding/controller.py [hunk 2]
```diff
@@ -62,19 +65,93 @@ def process(self, tasks: List[Task], **kwargs):
...
+        merged_task.worker_tag = self.WorkerTag.REWARD
+
+        # TODO: pack this logic
+        merged_task.max_tokens = 1
+        merged_task.return_context_logits = True
+
+        yield [merged_task]
+
+        assert merged_task.context_logits is not None
+        # TODO: consider running on cpu to not interrupt worker
```
Comment (WeiHaocheng) [inline tensorrt_llm/scaffolding/controller.py:140]: I think
"TODO: tokenizer should be run on worker."
would be clearer.

**Review by WeiHaocheng — APPROVED** [review]

File: examples/scaffolding/run_majority_vote_aime24.py
Comment (dc3671) [reply examples/scaffolding/run_majority_vote_aime24.py:1]: I add aime24 for this example will run part of aime24 dataset and is included in integration test, while other examples only run simple prompt.

File: examples/scaffolding/run_best_of_n_with_reward.py (hunk 1 above)
Comment (dc3671) [reply examples/scaffolding/run_best_of_n_with_reward.py:58]: I intentionally keep this line as a simple alternate prompt for debugging. The other prompt is a little long.

File: tests/integration/defs/test_e2e.py [hunk 3]
```diff
@@ -1992,7 +1992,7 @@ def test_ptp_scaffolding(llm_root, llm_venv, model_name, model_path):
     example_root = Path(os.path.join(llm_root, "examples", "scaffolding"))
     input_file = Path(os.path.join(example_root, "test.jsonl"))
     llm_venv.run_cmd([
-        str(example_root / "aime24_test.py"),
+        str(example_root / "run_majority_vote_aime24.py"),
         "--generation_dir",
```
Comment (WeiHaocheng) [inline tests/integration/defs/test_e2e.py:1996]: in this PR:
https://github.com/NVIDIA/TensorRT-LLM/pull/3731
I rename "--generation_dir" to "--model_dir", and change it as a requried field, please remember to change this line.

**Comment by dc3671** [comment]
/bot skip --comment "only tracked flaky tests failed"
